WordPress/src/main/java/org/wordpress/android/ui/sitecreation/SearchInputWithHeader.kt

```diff
@@ -0,0 +1,74 @@
+package org.wordpress.android.ui.sitecreation
+
+import android.support.v4.content.ContextCompat
+import android.support.v7.content.res.AppCompatResources
+import android.text.Editable
+import android.text.TextWatcher
+import android.view.View
+import android.view.ViewGroup
+import android.widget.EditText
+import android.widget.TextView
+import org.wordpress.android.R
+
+class SearchInputWithHeader(rootView: View, onClear: () -> Unit) {
+    private val headerLayout = rootView.findViewById<ViewGroup>(R.id.header_layout)
+    private val headerTitle = rootView.findViewById<TextView>(R.id.title)
+    private val headerSubtitle = rootView.findViewById<TextView>(R.id.subtitle)
+    private val searchInput = rootView.findViewById<EditText>(R.id.input)
+    private val progressBar = rootView.findViewById<View>(R.id.progress_bar)
+    private val clearAllButton = rootView.findViewById<View>(R.id.clear_all_btn)
+
+    var onTextChanged: ((String) -> Unit)? = null
+
+    init {
+        val context = rootView.context
+
+        val drawable = AppCompatResources.getDrawable(context, R.drawable.ic_search_white_24dp)
+        drawable?.setTint(ContextCompat.getColor(context, R.color.grey))
+        searchInput.setCompoundDrawablesRelativeWithIntrinsicBounds(drawable, null, null, null)
+        clearAllButton.background = drawable
+
+        val clearAllLayout = rootView.findViewById<View>(R.id.clear_all_layout)
+        clearAllLayout.setOnClickListener {
+            onClear()
+        }
+
+        searchInput.addTextChangedListener(object : TextWatcher {
+            override fun afterTextChanged(s: Editable?) {}
+            override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {}
+            override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {
+                onTextChanged?.invoke(s?.toString() ?: "")
+            }
+        })
+    }
+
+    fun setInputText(text: String) {
+        // If the text hasn't changed avoid triggering the text watcher
+        if (searchInput.text.toString() != text) {
+            searchInput.setText(text)
+        }
+    }
+
+    fun updateHeader(uiState: SiteCreationHeaderUiState?) {
+        uiState?.let {
```

Comment:
This was my mistake, but the animation needs to be set even when the uiState is null -> something like this

```suggestion
         val headerVisible = uiState != null
        if (!headerVisible && headerLayout.visibility == View.VISIBLE) {
            headerLayout.animate().translationY(-headerLayout.height.toFloat())
        } else if (headerVisible && headerLayout.visibility == View.GONE) {
            headerLayout.animate().translationY(0f)
        }
        uiState?.let {
            updateVisibility(headerLayout, true)
            headerTitle.text = uiState.title
            headerSubtitle.text = uiState.subtitle
        } ?: updateVisibility(headerLayout, false)
```

Reply:
Btw it also fixes #8751, so feel free to fix it in a different PR if you want to.

Reply:
Fixed in 0eb700331229db4b75ef58297c73d56b6281388c. I don't think `headerVisible` was the best name here because it sounds like it's representing the actual header's state. I renamed it to `headerShouldBeVisible` and implemented the change in the way you suggested.

P.S: I'd have actually preferred not to have a `null` state at all and change the `SiteCreationHeaderUiState` to be a sealed class with `Hidden` and `Visible` states. It's always best to avoid `null` when we can, because the meaning of it is not always super clear.

Reply:
I agree with you in general - I don't like using null as a value either. However, I dumped the solution with `Hidden` and `Visible` sealed classes mainly because it made the UiState way more complicated than it needed be and I realized it when I implemented the error states.

For example if you have a header - which can be visible and hidden -> but in the header is let's say subtitle which can again be visible or hidden. It just gets too complicated in my opinion.
  
I decided to do this exception and consider it as RecyclerView where the item is in the UiState and we render it or it's not there and we don't render it.

Reply:
In those type of cases what I like to do is for the base class to have the properties that we'll need, like `isVisible`, `isSubtitleVisible` etc and have them initiated from the classes that inherits it. The visibility thing is a common issue though, so I have been thinking about a way to streamline that, I just didn't have the time to flesh it out.

Reply:
Yep, that's exactly what I had -- an error base class and sub-classes for each error type -> but it felt overly complicated. Moreover it "forced" me to propagate error types to the UI which I didn't like. 
I agree this is not an ideal solution, but I still feel the price for its simplicity is worth it (comparing to what we had).